File: Чередников/medium 2192/2192.rs
use std::collections::HashSet;

fn dfs(input_edges: &Vec<Vec<usize>>, vertex: usize, store: &mut Vec<HashSet<i32>>) {
    if store[vertex].is_empty() == false {
        return;
    }

    if input_edges[vertex].len() == 0 {
        return;
    }
    for i in &input_edges[vertex] {
        store[vertex].insert(*i as i32);
        dfs(input_edges, *i, store);
        store[vertex] = store[vertex].union(&store[*i]).map(|&el| el).collect()
    }
}

impl Solution {
    pub fn get_ancestors(n: i32, edges: Vec<Vec<i32>>) -> Vec<Vec<i32>> {
        let n = n as usize;
        let mut graph = vec![vec![]; n];
        for edge in &edges {
            let out = edge[0] as usize;
            let input = edge[1] as usize;
            graph[input].push(out);
        }

        let mut store: Vec<HashSet<i32>> = Vec::new();
        for _ in 0..n {
            store.push(HashSet::new())
        }
        for i in 0..n {
            dfs(&graph, i, &mut store)
        }

        let ans = store
            .iter()
            .map(|el| {
                let mut tmp = el.iter().map(|&n| n).collect::<Vec<_>>();
                tmp.sort();
                tmp
            })
            .collect::<Vec<_>>();

        return ans;
    }
}

File: Штоколова/2192.py
class Solution:
    def getAncestors(self, n: int, edges: List[List[int]]) -> List[List[int]]:

        res = []
        for i in range(n):
            res.append([])
        
        if len(edges) == 0:
            return res

        edges.sort()

        num = [0]*n 
        j = 0
        for i in range(0, n): 
          num[i] = j
          while (j < len(edges) and edges[j][0] == i): 
            j += 1
          if edges[j-1][0] != i:
            num[i] = -1 

        def dfs(current_v: int, v:int):
          i = num[current_v]

          while i < len(edges) and edges[i][0] == current_v:
            neighbor_v = edges[i][1]
            size_v = len(res[neighbor_v]) - 1
            if size_v >= 0 and res[neighbor_v][size_v] != v:
              res[neighbor_v].append(v)
              if num[neighbor_v] != -1:
               dfs(neighbor_v, v)
            elif size_v < 0:
              res[neighbor_v].append(v)
        
              if num[neighbor_v] != -1:
                dfs(neighbor_v, v)
            i += 1
            
        for k in range(n):
          if num[k] != -1: #and not use[k]:
            dfs(k, k) 
          

        return res 
